### PR TITLE
Optimized/Rebuild `LevelDB` Native for `LevelDBStore`

### PR DESCRIPTION
**Pull request by cschuchardt88**
# Change Log
- Optimized `LevelDBStore`
- Code Cleanup
- Rebuild

image

### CODE REVIEW

**Comment by cschuchardt88** [comment]
@cschuchardt88 @Jim8y @vncoelho @superboyiii @chenzhitong @AnnaShaleva 
Review Please

**Review by vncoelho — COMMENTED** [review]
@cschuchardt88 ,I would like to see a benchmark on this.
This is not the first time this part is being refactoring by you, right? Or is this continuation of previous commit?

**Comment by cschuchardt88** [comment]
@vncoelho This is refactoring everything. Im sure there is performance benefits. But that isn't the point of this PR. This PR fixes `dotnet` management of the `leveldb` objects from `marshal` objects. Also you have access the `leveldb` classes for easy accessibly.

Example
---
```csharp
var options = new Options { CreateIfMissing = true };
using (var db = new DB(options, path))
{
    db.Put("NA", "Na");

    using(var batch = new WriteBatch())
    {
        batch.Delete("NA")
             .Put("Tampa", "Green")
             .Put("London", "red")
             .Put("New York", "blue");
        db.Write(batch);
    }
}
```

File: src/LevelDBStore/DB.cs [hunk 1]
```diff
@@ -0,0 +1,459 @@
...
+        /// </summary>
+        public DB(Options options, string name, Encoding encoding)
+        {
+            this._cache = options.Cache;
+            this._infoLog = options.InfoLog;
+            this._comparator = options.Comparator;
+            this.Handle = LevelDBInterop.leveldb_open(options.Handle, name, out var error);
+            this._encoding = encoding;
+
+            Throw(error, msg => new UnauthorizedAccessException(msg));
```
Comment (shargon) [inline src/LevelDBStore/DB.cs:72]: Previously we call `FilterPolicy = Native.leveldb_filterpolicy_create_bloom(15)`

File: src/LevelDBStore/ReadOptions.cs [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+        /// If "snapshot" is provides, read as of the supplied snapshot
+        /// (which must belong to the DB that is being read and which must
+        /// not have been released).  
+        /// If "snapshot" is not set, use an implicit
+        /// snapshot of the state at the beginning of this read operation.
+        /// </summary>
+        public SnapShot Snapshot
+        {
+            set { LevelDBInterop.leveldb_readoptions_set_snapshot(this.Handle, value.Handle); }
+        }
```
Comment (shargon) [inline src/LevelDBStore/ReadOptions.cs:55]: No getters?

File: src/LevelDBStore/LeastRecentlyUsedSet.cs [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+// LeastRecentlyUsedSet.cs file belongs to the neo project and is free
+// software distributed under the MIT software license, see the
+// accompanying file LICENSE in the main directory of the
+// repository or http://www.opensource.org/licenses/mit-license.php
+// for more details.
+//
+// Redistribution and use in source and binary forms with or without
+// modifications are permitted.
+
+#if false
```
Comment (shargon) [inline src/LevelDBStore/LeastRecentlyUsedSet.cs:12]: ?

**Review by shargon — COMMENTED** [review]
Reviewed 17/36

File: src/LevelDBStore/LevelDBHandle.cs [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+                if (disposing)
+                {
+                    FreeManagedObjects();
+                }
+                if (this.Handle != IntPtr.Zero)
+                {
+                    FreeUnManagedObjects();
+                    this.Handle = IntPtr.Zero;
+                }
+                _disposed = true;
```
[the remark this answers is not in the document]
Comment (cschuchardt88) [reply src/LevelDBStore/LevelDBHandle.cs:52]: `_disposed` is saying have already `disposed` objects.

File: src/LevelDBStore/DB.cs (hunk 1 above)
Comment (cschuchardt88) [reply src/LevelDBStore/DB.cs:72]: What does that do or even mean for `leveldb`?

File: src/LevelDBStore/ReadOptions.cs (hunk 2 above)
Comment (cschuchardt88) [reply src/LevelDBStore/ReadOptions.cs:55]: There is no need for `get` `leveldb` doesn't support getting the options values.

File: src/LevelDBStore/LeastRecentlyUsedSet.cs (hunk 3 above)
Comment (cschuchardt88) [reply src/LevelDBStore/LeastRecentlyUsedSet.cs:12]: Its disabled.

Comment (shargon) [reply src/LevelDBStore/LeastRecentlyUsedSet.cs:12]: Then we can remove it

File: src/LevelDBStore/ReadOptions.cs (hunk 2 above)
Comment (shargon) [reply src/LevelDBStore/ReadOptions.cs:55]: Seems weird only set property, then it's better to use a init method

File: src/LevelDBStore/LevelDBHandle.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (shargon) [reply src/LevelDBStore/LevelDBHandle.cs:52]: Prevents double layout in multiple threads.

File: src/LevelDBStore/ReadOptions.cs (hunk 2 above)
Comment (cschuchardt88) [reply src/LevelDBStore/ReadOptions.cs:55]: No, can still do for example:
---
```csharp
var readOptions = new ReadOptions();
readOptions.SnapShot = snapshot;
```

File: src/LevelDBStore/LevelDBHandle.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (cschuchardt88) [reply src/LevelDBStore/LevelDBHandle.cs:52]: Moving above all that doesn't do anything either way. Only difference would be you set the `_disposed` variable early.

File: src/LevelDBStore/DB.cs (hunk 1 above)
Comment (cschuchardt88) [reply src/LevelDBStore/DB.cs:72]: Fixed add it, and it set bloom to `10` should need no more than `10` bits.

**Comment by cschuchardt88** [comment]
@shargon @Jim8y review please

**Review by shargon — COMMENTED** [review]
@cschuchardt88 You never use `this.` and the code have a lot of `this.`. Is this code yours?

File: src/LevelDBStore/WriteOptions.cs [hunk 5]
```diff
@@ -16,38 +16,40 @@ namespace LevelDB
...
-            set { LevelDBInterop.leveldb_writeoptions_set_sync(this.Handle, value ? (byte)1 : (byte)0); }
-        }
+            Handle = LevelDBInterop.leveldb_writeoptions_create();
 
+            if (sync != null)
+            {
+                // Set sync mode
+                LevelDBInterop.leveldb_writeoptions_set_sync(Handle, sync.Value ? (byte)1 : (byte)0);
+            }
+        }
```
Comment (cschuchardt88) [inline src/LevelDBStore/WriteOptions.cs]: @shargon why did you change this?

File: src/LevelDBStore/Comparator.cs [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+                        var @this = GCHandle.FromIntPtr(gCHandleThis).Target as Inner;
+                        return @this.NameValue;
+                    };
+
+            private Func<NativeArray, NativeArray, int> _cmp;
+            private GCHandle _namePinned;
+
+            public IntPtr Init(string name, Func<NativeArray, NativeArray, int> cmp)
+            {
+                // TODO: Complete member initialization
```
Comment (shargon) [inline src/LevelDBStore/Comparator.cs:69]: We should finish the code

File: src/LevelDBStore/ReadOptions.cs [hunk 7]
```diff
@@ -16,47 +16,53 @@ namespace LevelDB
...
+            {
+                // set verify checksums
+                LevelDBInterop.leveldb_readoptions_set_verify_checksums(Handle, verifyCheckSums.Value ? (byte)1 : (byte)0);
+            }
+
+            if (fillCache != null)
+            {
+                // set fill cache
+                LevelDBInterop.leveldb_readoptions_set_fill_cache(Handle, fillCache.Value ? (byte)1 : (byte)0);
+            }
```
Comment (cschuchardt88) [inline src/LevelDBStore/ReadOptions.cs]: @shargon  why change this?

**Review by cschuchardt88 — COMMENTED** [review]
@shargon to answer your question I was keeping consistent with existing code in that library.

**Review by AnnaShaleva — COMMENTED** [review]
This PR contains a massive set of changes, it's hard to review and some mistake may be done occasionally. There should be a way to split this PR into multiple granular PRs, each PR for its own optimisation. It also would be nice to distinguish optimisations and code refactoring. 

Also, every optimisation should be confirmed by benchmarks, otherwise it's controversial. I'm sure there is a way of C# code profiling. Take a look at https://github.com/nspcc-dev/neo-go/pull/2330/commits/cd42b8b20ca463400678ee52dfd2e9d6b9620fa5, https://github.com/nspcc-dev/neo-go/pull/2659/commits/9712be78fd01d056bcb1ea7474238ebc23f30a03, https://github.com/nspcc-dev/neo-go/pull/2290/commits/ffeb3b8473ee49dc4b7516add1f291139958b720 and other optimisation-related PRs in NeoGo.

There shouldn't be TODOs in the resulting code, since every TODO is an issue.

Also, have you tested the resulting code against the current mainnet/testnet? Is the state compatible?

File: src/LevelDBStore/Comparator.cs (hunk 6 above)
Comment (cschuchardt88) [reply src/LevelDBStore/Comparator.cs:69]: It's there if you need it.

@shargon i reverted all your changes. Please don't make changes to my code directly unless you have good reason. You change my code all the time to leave it broken or add bugs and you don't understand the reason behind the design.

**Comment by cschuchardt88** [comment]
@AnnaShaleva 
>Also, have you tested the resulting code against the current mainnet/testnet? Is the state compatible?

Yes

This adds the full features of the leveldb library.
